File: elixir/lib/advent_of_code/day_10.ex
defmodule AdventOfCode.Day10 do
  def count(n, [], pre), do: 0

  def count(n, [hd | tl], pre) do
    if hd - pre == n do
      1 + count(n, tl, hd)
    else
      0 + count(n, tl, hd)
    end
  end

  def ways([], _, paths), do: paths

  def ways([hd | tl], ports, paths) do
    children = Map.get(ports, hd)

    new_paths =
      Enum.reduce(children, paths, fn child, acc ->
        child_val = Map.get(acc, child)
        Map.update!(acc, hd, &(&1 + child_val))
      end)

    ways(tl, ports, new_paths)
  end

  def part1(input) do
    count(1, input, 0) * (count(3, input, 0) + 1)
  end

  def part2(input) do
    maximum = Enum.max(input) + 3
    list = [0 | [maximum | input]] |> Enum.sort() |> input_to_map
    topo_sort = list |> Enum.map(fn {x, _} -> x end) |> Enum.sort() |> Enum.reverse()

    number_map =
      topo_sort
      |> Enum.zip(for _ <- 0..length(topo_sort), do: 0)
      |> Map.new()
      |> Map.update!(maximum, fn _ -> 1 end)

    Map.get(ways(topo_sort, list, number_map), 0)
  end

  def input_to_map(list) do
    list
    |> Enum.map(fn x -> {x, 1..3 |> Enum.map(&(&1 + x)) |> Enum.filter(&(&1 in list))} end)
    |> Map.new()
  end

  def parse_input(txt) do
    txt |> String.split("\n") |> Enum.map(&String.to_integer(&1)) |> Enum.sort()
  end
end

File: haskell/src/Day10.hs
module Day10 where

import Data.List
import qualified Data.Map as M
import Data.Maybe

count1s :: [Int] -> Int -> Int
count1s [] _ = 0
count1s (x : xs) pre = if x - pre == 1 then 1 + count1s xs x else 0 + count1s xs x

count3s :: [Int] -> Int -> Int
count3s [] _ = 0
count3s (x : xs) pre = if x - pre == 3 then 1 + count3s xs x else 0 + count3s xs x

ways :: [Int] -> [(Int, [Int])] -> M.Map Int Int -> M.Map Int Int
ways [] _ paths = paths
ways (x : xs) ports paths =
  let children = fromMaybe [] . lookup x $ ports
      newPaths =
        foldl
          ( \acc child ->
              let Just childVal = M.lookup child acc
               in M.update (\val -> Just (val + childVal)) x acc
          )
          paths
          children
   in ways xs ports newPaths

part1 :: [Int] -> Int
part1 input =
  let ones = count1s input 0
      threes = count3s input 0 + 1
   in ones * threes

part2 :: [Int] -> Maybe Int
part2 input =
  let maxL = maximum input + 3
      list = inputToMap . sort $ 0 : maxL : input
      topoSort = reverse . sort . map fst $ list
      numberMap = M.update (\_ -> Just 1) maxL $ M.fromList $ zip topoSort (repeat 0)
   in M.lookup 0 $ ways topoSort list numberMap

inputToMap :: [Int] -> [(Int, [Int])]
inputToMap [] = []
inputToMap (x : xs) = (x, filter (`elem` xs) . map (+ x) $ [1 .. 3]) : inputToMap xs

parseInput :: String -> [Int]
parseInput = sort . map read . lines
